### PR TITLE
Raise nicer errors when legislation is invalid

### PR DESCRIPTION
**Pull request by michelbl**
Connected to #474 

## Technical improvement

Improve error message during parameters validation, using lxml with a XML schema. No breaking change for users.

### CODE REVIEW

**Comment by michelbl** [comment]
@fpagnoux Beware ! This PR may break the web API.

**Comment by fpagnoux** [comment]
How can it break the API if there is no breaking change :D ?
Which change do you think may have an impact on the API ?

File: openfisca_core/legislationsxml.py [hunk 1]
```diff
@@ -3,926 +3,198 @@
...
-    return message
+    with open(filename_xml_schema, 'r') as f:
+        xmlschema_doc = etree.parse(f)
+        xmlschema = etree.XMLSchema(xmlschema_doc)
 
+    return xmlschema
 
-# Level 1 converters
 
+def parse_and_validation_xml(xmlschema, legislation_xml_info_list):
```
Comment (fpagnoux) [inline openfisca_core/legislationsxml.py]: Prefer verbs for methods. `parse_and_validate_xml` ?

File: openfisca_core/legislationsxml.py [hunk 2]
```diff
@@ -3,926 +3,198 @@
...
+
+    return merged_json
+
+
+def load_legislation(legislation_xml_info_list):
+    xmlschema = load_xml_schema()
+
+    xml_trees = parse_and_validation_xml(xmlschema, legislation_xml_info_list)
+
+    name_list, json_list = transform(xml_trees)
```
Comment (fpagnoux) [inline openfisca_core/legislationsxml.py]: `transform` is a bit vague. What kind of transformation is done by this function ?
Does it mostly convert XML to JSON ?

File: openfisca_core/legislation.xsd [hunk 3]
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
+
+    <xs:element name="NODE" type="nodeType"/>
+
+    <xs:group name="piecewiseConstant">
```
Comment (fpagnoux) [inline openfisca_core/legislation.xsd]: What is a `piecewiseConstant ` ? I've seen this term a lot in this PR, but I don't really get what it represents.

File: tests/core/parameter_validation/test_parameter_validation.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
```
Comment (michelbl) [inline tests/core/parameter_validation/test_parameter_validation.py]: thanks :blush:

File: openfisca_core/legislationsxml.py (hunk 1 above)
Comment (michelbl) [inline openfisca_core/legislationsxml.py]: ok

File: openfisca_core/legislationsxml.py (hunk 2 above)
Comment (michelbl) [inline openfisca_core/legislationsxml.py]: It converts ElementTree to JSON, so I suggest `transform_etree_to_json_root` / `transform_etree_to_json_recursive`.

File: openfisca_core/legislation.xsd (hunk 3 above)
Comment (michelbl) [inline openfisca_core/legislation.xsd]: All numeric parameters in the XML are piecewise constant functions of the time (or step functions), cf https://en.wikipedia.org/wiki/Step_function. Maybe "stepFunction" would be clearer ?

**Comment by michelbl** [comment]
screenshot from 2017-05-15 15-54-53
Wow don't be angry at me @fpagnoux !

If the tests have been moved, then there should be no problem.

**Comment by fpagnoux** [comment]
The SGMAP internet is angry, not me ;)

File: openfisca_core/legislation.xsd (hunk 3 above)
Comment (fpagnoux) [inline openfisca_core/legislation.xsd]: Ok, I get it. So for instance this is a `piecewiseConstant `:
```xml
      <END deb="2012-01-01" />
      <VALUE deb="2009-01-01" valeur="0.6" />
      <VALUE deb="2008-01-01" valeur="0.5" />
```
Not sure this is the best way to describe it in the context of OpenFisca. Some suggestions : 
- `valuesHistory`
- `datedValues`
- `parameterHistory`

Comment (michelbl) [inline openfisca_core/legislation.xsd]: `valuesHistory` is good to me

File: openfisca_core/legislationsxml.py [hunk 5]
```diff
@@ -3,926 +3,198 @@
...
-        transform_value_xml_json_to_json(item, float)
-        for item in values_holder_xml_json['VALUE']
-        ]
-
-    if 'END' in values_holder_xml_json:
-        ends = [
-            transform_end_xml_json_to_json(item)
-            for item in values_holder_xml_json['END']
-            ]
+            raise ValueError('Unknown attribute {} : {}'.format(key, value))
```
Comment (cbenz) [inline openfisca_core/legislationsxml.py]: Python 2 warning: string in exception must be 'utf-8' encoded, so:
```py
raise ValueError(u'Unknown attribute {}: {}'.format(key, value).encode('utf-8'))
```

(There is hope: in Python 3 this constraint disappears.)

Comment (cbenz) [inline openfisca_core/legislationsxml.py]: No ` ` between `:` in english

File: openfisca_core/legislationsxml.py [hunk 6]
```diff
@@ -3,926 +3,198 @@
...
+            if child.tag == 'SEUIL':
+                json_node['threshold'] = new_child
+            elif child.tag == 'TAUX':
+                json_node['rate'] = new_child
+            elif child.tag == 'ASSIETTE':
+                json_node['base'] = new_child
+            elif child.tag == 'MONTANT':
+                json_node['amount'] = new_child
+            else:
+                raise ValueError('Unknown TRANCHE child {}'.format(child.tag))
```
Comment (cbenz) [inline openfisca_core/legislationsxml.py]: > Same remark about utf-8 messages, but *here* I know that `child.tag` will be ASCII-only so won't break

File: openfisca_core/legislationsxml.py (hunk 5 above)
Comment (michelbl) [inline openfisca_core/legislationsxml.py]: ok

Comment (michelbl) [inline openfisca_core/legislationsxml.py]: ok

Comment (michelbl) [inline openfisca_core/legislationsxml.py]: ok

File: openfisca_core/legislationsxml.py (hunk 6 above)
Comment (michelbl) [inline openfisca_core/legislationsxml.py]: ok

File: check-version-bump.sh [hunk 7]
```diff
@@ -7,7 +7,7 @@ then
     if git rev-parse $current_version
     then
         set +x
-        echo "Version $version already exists. Please update version number in setup.py before merging this branch into master."
+        echo "Version $current_version already exists. Please update version number in setup.py before merging this branch into master."
```
Comment (cbenz) [inline check-version-bump.sh]: Nice

File: openfisca_core/legislation.xsd [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        <xs:sequence>
+            <xs:choice minOccurs="0" maxOccurs="unbounded">
+                <xs:element name="NODE" type="nodeType"/>
+                <xs:element name="CODE" type="codeType"/>
+                <xs:element name="BAREME" type="baremeType"/>
+            </xs:choice>
+        </xs:sequence>
+        <xs:attribute name="code" type="xs:string" use="required"/>
+        <xs:attribute name="description" type="xs:string"/>
+        <xs:attribute name="origin" type="xs:string"/>
```
Comment (MattiSG) [inline openfisca_core/legislation.xsd]: > In order to increase readability, always set the fixed-length attributes at the beginning 🙂

File: tests/core/parameter_validation/test_parameter_validation.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+from lxml.etree import XMLSyntaxError
+
+from openfisca_core.taxbenefitsystems import TaxBenefitSystem
+from openfisca_dummy_country.entities import entities
+from openfisca_dummy_country.scenarios import Scenario
+
+
+BASE_DIR = os.path.dirname(os.path.abspath(__file__))
+
+year = 2016
```
Comment (MattiSG) [inline tests/core/parameter_validation/test_parameter_validation.py]: If this is a constant, CAPITALS should be used :wink:

File: tests/core/parameter_validation/test_parameter_validation.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        TaxBenefitSystem.__init__(self, entities)
+        self.Scenario = Scenario
+        self.add_legislation_params(path_to_xml)
+
+
+@raises(XMLSyntaxError)
+def test_malformed():
+    path_to_xml = os.path.join(BASE_DIR, 'malformed.xml')
+    tbf = TestTaxBenefitSystem(path_to_xml)
+    tbf.compute_legislation()
```
Comment (MattiSG) [inline tests/core/parameter_validation/test_parameter_validation.py]: > I usually like to call my [SUT](https://en.wikipedia.org/wiki/System_under_test) “subject” in order to improve test readability 🙂

File: tests/core/parameter_validation/test_parameter_validation.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+def test_unknown_node():
+    path_to_xml = os.path.join(BASE_DIR, 'unknown_node.xml')
+    tbf = TestTaxBenefitSystem(path_to_xml)
+    try:
+        tbf.compute_legislation()
+    except ValueError as e:
+        content = str(e)
+        assert len(content) < 1000
+        for keyword in {'unknown_node.xml:4', 'VALEU', 'not expected', 'VALUE'}:
+            assert keyword in content
```
Comment (MattiSG) [inline tests/core/parameter_validation/test_parameter_validation.py]: :clap:

**Pull request merged**